### ISSUE TITLE
Resolution of local id is made relative to id of a sibling not parent

### ISSUE THREAD

**Issue by maciasello** [ajv-validator/ajv#170]
When using schema provided below the validation fails with error although IMHO it is correct ("file" property is a sibling to "title" so "title"'s id should not affect its scope):

> ajv compile -s schema.json
> schema schema.json is invalid
> error: can't resolve reference #/definitions/file-entry from id http://example.com/title

Schema:

``` json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "id": "http://example.com/base",
  "type": "object",
  "properties": {
    "title": {
      "id": "http://example.com/title",
      "type": "string"
    },
    "file": {
      "$ref": "#/definitions/file-entry"
    }
  },
  "definitions": {
    "file-entry": {
      "type": "string"
    }
  }
}
```

**Comment by epoberezkin** [comment]
Indeed: https://tonicdev.com/esp/571a223b2449f212002e0713

**Comment by epoberezkin** [comment]
@maciasello you are right, it's a bug. Thanks for reporting. I would appreciate if you could investigate further or to submit a failing test - I will try to fix soon anyway.

**Comment by epoberezkin** [comment]
Keywords where this issue with leaking change of resolution scope should be fixed:
- [x] properties
- [x] items
- [x] allOf
- [x] anyOf
- [x] dependencies
- [x] oneOf
- [x] switch

**Comment by epoberezkin** [comment]
fixed in 4.0.4

**Comment by maciasello** [comment]
That was quick! Thanks! Confirmed working in my case!
